packages/loopring_v3/contracts/aux/bridge/Bridge.sol

```diff
@@ -0,0 +1,902 @@
+                header & 0xffff == (uint(ExchangeData.TransactionType.WITHDRAWAL) << 8) | 1 &&
+                // withdrawal.tokenID == token.tokenID &&
+                // withdrawal.amount == token.amount &&
+                // withdrawal.fee == 0,
+                packedData & 0xffffffffffffffffffffffffffff0000ffff == (uint(token.tokenID) << 128) | (token.amount << 32) &&
+                onchainDataHash == dataHash,
+                "INVALID_BRIDGE_WITHDRAWAL_TX_DATA"
+            );
+
+            ctx.txsDataPtr += ExchangeData.TX_DATA_AVAILABILITY_SIZE;
+        }
+    }
+
+    function _storeTransfers(
+        bytes memory transfers,
+        address from
+        )
+        internal
+    {
+        uint batchID = batchIDGenerator++;
+
+        // Store transfers to distribute at a later time
+        bytes32 hash = _hashTransferBatch(transfers);
+        require(pendingTransfers[batchID][hash] == 0, "DUPLICATE_BATCH");
+        pendingTransfers[batchID][hash] = block.timestamp;
+
+        // Log transfers to do
+        emit Transfers(batchID, transfers, from);
+    }
+
+    function _deposit(
+        address from,
+        address token,
+        uint96  amount
+        )
+        internal
+    {
+        if (amount == 0) {
+            return;
+        }
+
+        if (from == address(this) && token != address(0)) {
+            ERC20(token).approve(address(depositContract), amount);
+        }
+        // Do the token transfer directly to the exchange
+        uint ethValue = (token == address(0)) ? amount : 0;
+        exchange.deposit{value: ethValue}(from, address(this), token, amount, new bytes(0));
+    }
+
+    function _connectorCall(
+        Context          memory   ctx,
+        ConnectorCalls   calldata connectorCalls,
+        uint                      n,
+        ConnectorCalls[] calldata allConnectorCalls
+        )
+        internal
+        returns (BridgeTransfer[] memory transfers)
+    {
+        require(connectorCalls.connector != address(this), "INVALID_CONNECTOR");
+        require(trustedConnectors[connectorCalls.connector], "ONLY_TRUSTED_CONNECTORS_SUPPORTED");
+
+        // Check if the minimum amount of gas required is achieved
+        bytes memory txData = _getConnectorCallData(ctx, IBridgeConnector.getMinGasLimit.selector, allConnectorCalls, n);
+        (bool success, bytes memory returnData) = connectorCalls.connector.fastCall(GAS_LIMIT_CHECK_GAS_LIMIT, 0, txData);
+        if (success) {
+            require(connectorCalls.gasLimit >= abi.decode(returnData, (uint)), "GAS_LIMIT_TOO_LOW");
+        } else {
+            // If the call failed for some reason just continue.
+        }
+
+        // Execute the logic using a delegate so no extra transfers are needed
+        txData = _getConnectorCallData(ctx,IBridgeConnector.processCalls.selector, allConnectorCalls, n);
+        (success, returnData) = connectorCalls.connector.fastDelegatecall(connectorCalls.gasLimit, txData);
```

Comment:
shall we have something like this:
```if (trusted) connector.fastDelegateCall() else connector.fastCall()```

Reply:
Non-trusted connectors are a bit more different then that because for a normal call to work we would first have to transfer the necessary funds the connector needs to that connector contract. And also some other differences like not being able to trust any data returned from the contract so non-trusted connectors have to do a batchDeposit inside the connector call if needed.

Trusted connectors are easier/cheaper because we can just a delegate call and all code is executed as if it was from the Bridge contract. And it's the Bridge contract that contains all user funds necessary for the connector calls.